bundles/mapping/mapmodule/service/map.layer.js

```diff
@@ -329,10 +344,191 @@ Oskari.clazz.define('Oskari.mapframework.service.MapLayerService',
+                for (var i = 0; i < groupLayers.length; i++) {
+                    var layer = groupLayers[i];
+                    if (layer.id === layerId) {
+                        founded = i;
+                        break;
+                    }
+                }
+                return founded;
+            };
+
+
+            // update or insert
+            if (!deleteLayer) {
+                var layerIndex = null;
+
+                for (var i = 0; i < me._layerGroups.length; i++) {
+                    group = me._layerGroups[i];
+
+                    // Check if layer is in main groups
+                    layerIndex = getLayerIndexInArray(group.layers);
+                    if (layerIndex >= 0) {
+                        if (deleteLayer) {
+                            group.layers.splice(layerIndex, 1);
+                        } else {
+                            group.layers[layerIndex] = newLayerConf;
+                        }
+                    }
+
+                    // check subgroups
+                    for (var j = 0; j < group.groups.length; j++) {
+                        var subgroup = group.groups[j];
+
+                        layerIndex = getLayerIndexInArray(subgroup.layers);
+                        if (layerIndex >= 0) {
+                            if (deleteLayer) {
+                                group.groups[j].layers.splice(layerIndex, 1);
+                            } else {
+                                group.groups[j].layers[layerIndex] = newLayerConf;
+                            }
+                        }
+                        // check subgroup subgroups
+                        if (subgroup.groups) {
+                            for (var k = 0; k < subgroup.groups.length; k++) {
+                                var subgroupSubgroup = subgroup.groups[k];
+                                layerIndex = getLayerIndexInArray(subgroupSubgroup.layers);
+                                if (layerIndex >= 0) {
+                                    if (deleteLayer) {
+                                        group.groups[j].groups[k].layers.splice(layerIndex, 1);
+                                    } else {
+                                        group.groups[j].groups[k].layers[layerIndex] = newLayerConf;
+                                    }
+                                }
+                            }
+                        }
+
+                        if (layerIndex >= 0) {
+                            break;
+                        }
+                    }
+
+                    if (layerIndex >= 0) {
+                        break;
+                    }
+
+                }
+
+
+                // Finally check at layer has all groups in dom
+                newLayerConf.groups.forEach(function(group) {
+                    var groupConf = me.getAllLayerGroups(group.id);
+                    var isInGroup = jQuery.grep(me.getAllLayerGroups(group.id).layers, function(layer) {
```

Comment:
You are using both Array.filter and jQuery.grep in this file. Can you change this to use filter?